keps/sig-scheduling/20180409-scheduling-framework.md

```diff
@@ -264,6 +266,23 @@ state, it will either trigger [Un-reserve](#un-reserve) plugins (on failure) or
 
 *Note: This concept used to be referred to as "assume".*
 
+### VictimsSelection
+
+The plugin provides an extension point for selecting the victim pods for preemption.  Only one preempt plugin may be enabled at a time.  
+
+The selection process are performed in cooperating with scheduler and preempt plugin as follows.
+
+1. Scheduler firstly filters _potential victim nodes_ that satisfies the following two conditions.  It contains _potential victim pods_ that are allowed to be preempted by the preemptor pod's `spec.preemptionPolicy`. For example, if the preemptor pods' `spec.preemptionPolicy` is `Never`,  no victims are selected for any nodes.  The other is the preemptor pod can be schedulable when all the potential victim pods are assumed to be preempted.
+1. For each potential victim node, scheduler filters several pods as _victim candidate pods_ from potential victim pods with the plugin's help (see `SelectVictimCandidatesOnNode`).
+1. Scheduler calls scheduler extenders, if configured, to filter victim candidate pods.
+1. Scheduler picks one node from the potential victim nodes for preemption with the plugin's help (see `PickOneNodeForPreemption`).
+
+Thus, the plugin can provide two methods to customize behavior of victim pods selection.
+
+- `SelectVictimCandidatesOnNode`: The method receives a potential victim node and its potential victim pods.  And, the method returns victim candidate pods.  Please note the plugin is responsible that the returned victims can create enough room in the node for the preemptor pod.  That means, even if it can't create enough space for the preemptor pod, the scheduling cycle will not abort. Scheduler also passes a predicate function which accept victim pods and returns they can make enough room for the preemptor pod or not.  You could use this predicate function iteratively to find minimal victim pods.
```

Comment:
I 'm wondering how to feed `podPassesFiltersOnNode` in the generic scheduler to this method.  I think we can feed the function as an argument of `SelectVictimCandidatesOnNode`.

Reply:
> the function could be part of the framework handle instead (this might require a non-trivial effort).

When I drafted this, I thought so.   However, as you commented, `framework` would be difficult to depend on `genericScheduler` instance for now.  So, I prefer to pass the function as an argument. What do you think?
 
> What if the extension point just returns an order to be used by the core preemption algorithm?

Do you mean `func SelectVictimCandidatesOnNode(..., potentialVictims []*v1.Pod) ([]int, *Status)` ??  this sounds nice.

Reply:
> However, as you commented, framework instance would be difficult to depend on genericScheduler instance in the current implementation

Rather `podPassesFiltersOnNode` would move to the framework. I think the major problem there are extenders. That said, I don't quite like this, but I find it better than passing a function.

> Do you mean func SelectVictimCandidatesOnNode(..., potentialVictims []*v1.Pod) ([]int, *Status) ?? this sounds nice.

Not sure what you mean by the returned `[]int`. But I meant something like:

```func SortVictimCandidatesOnNode(..., potentialVictims []*v1.Pod) ([]*v1.Pod, *Status)```

Essentially replacing https://github.com/kubernetes/kubernetes/blob/51b2d02717399e4ef1c207ea5d913a8bdb7bd994/pkg/scheduler/core/generic_scheduler.go#L994, but it could also filter out some victims

Reply:
> I guess the function could be part of the framework handle instead (this might require a non-trivial effort).

+1. Maybe as well as `filterPodsWithPDBViolation()`.

Regarding the Sorting thing, I think it should be an internal piece inside the plugin's implementation, instead of a public interface. With that said, my idea is: expose `selectNodesForPreemption()` (caller of `selectVictimsOnNode()`) as an interface, with all necessary core algorithm as arguments in framework handle. With this, we can support plugins to implement an advanced strategy to preempt Pods among different Nodes.

Reply:
>  find it better than passing a function.

I agree.

> `func SortVictimCandidatesOnNode(..., potentialVictims []*v1.Pod) ([]*v1.Pod, *Status)`

Thanks for clarifying.  I understood your idea.   

I think  Wei's `PreemptionStrategy` based approach is generic enough.  And as Wei commented on https://github.com/kubernetes/enhancements/pull/1398/files#r394728508, this can also allow preempting multiple victim nodes (I think this is a very aggressive use case, though).

> With that said, my idea is: expose `selectNodesForPreemption()` (caller of `selectVictimsOnNode()`) as an interface, with all necessary core algorithm as arguments in framework handle. With this, we can support plugins to implement an advanced strategy to preempt Pods among different Nodes.

That sounds nice 👍

Reply:
In general, my strategy while reviewing this is to put as much logic as possible in the core scheduler. I believe that what I have proposed is the minimal enough to cover @everpeace use case.

But now @Huang-Wei is also proposing support for muti-node preemption, which my proposal wouldn't satisfy. However, I would like to know if we know of use cases of this. Even for something like pod spreading: we know that's a use case, but is it something that we plan to solve or, even, that CAN be solved without introducing a significant latency? That problem is NP (it already is for a single node).

Reply:
> But now @Huang-Wei is also proposing support for muti-node preemption, which my proposal wouldn't satisfy. However, I would like to know if we know of use cases of this.

I think this is one use cases of this :  PodGroup with high-priority preempt other PodGroup with low-priority in Coscheduling when resources are tight. PodGroup with low-priority maybe runs on muti-node. @alculquicondor

Reply:
The question remains: can we solve that problem efficiently?